source/common/http/conn_manager_impl.cc

```diff
@@ -445,8 +440,10 @@ const Network::Connection* ConnectionManagerImpl::ActiveStream::connection() {
 }
 
 void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers, bool end_stream) {
-  maybeEndDecode(end_stream);
   request_headers_ = std::move(headers);
+  lazilyCreateFilterChain();
```

Comment:
Q: Why do we need the lazy init functionality here? IIRC this function will only be called once? Also, I think we will eventually want to do this lower down in this function once we decide if it's WS or not? Should we position it in the final place now? Perhaps above here:
https://github.com/envoyproxy/envoy/blob/95d6286ad586bf0faea3cd0a8d056dcec87a9216/source/common/http/conn_manager_impl.cc#L696

(The downside of putting it there is any local replies in this function would not go through encoding filters... which I guess is an issue, but it seems like if we instantiate here we would need to blow away and recreate for WS which seems worse? Thoughts?)

Reply:
So for local replies, there's a TODO to start encoding from the last filter which saw the request.  If in the long run we want filters to only see the response if they saw the request I think this is fine.  The other option is we can lazily create the chain on any local reply for consistency until we fix that TODO.  I'd assumed we'd have other entry points which is one reason I made it a function :-)

I'd prefer we create the filter chain early on in the function.  I'd planned on lazilyCreateFilterChain being refactored to look at request_headers_ (and config, to see if the new path is configured) then deciding if it should create websocket or http filter chain, which is why I stuck it immediately after we latch request_headers_.  Happy to move it if you think that's better but I would like to land it without the refactor just to keep things as low risk as possible :-)

Reply:
If you think you are going to put inspection logic in the creation function, I think it's fine to keep it where it is. My preference though is to remove the lazy creation logic though until we actually need it? So maybe just a function callout with a comment about lazy creation, etc.?

Reply:
> Sorry, are you just asking that we create the filter chain when this function is called, kill off the boolean, and just rename it to createFilterChain?

Yes that's what I mean. Unless I'm missing something, in the current code the lazy creation checking logic will never be exercised (we will only call the creation function once).